dowhy/causal_identifier.py

```diff
@@ -244,51 +252,59 @@ def identify_backdoor(self, treatment_name, outcome_name):
             backdoor_sets.append({
                 'backdoor_set':empty_set,
                 'num_paths_blocked_by_observed_nodes': check["num_paths_blocked_by_observed_nodes"]})
+            # If the method is `minimal-adjustment`, return the empty set right away.
+            if method_name == CausalIdentifier.BACKDOOR_MIN:
+                return backdoor_sets
+
         # Second, checking for all other sets of variables
         eligible_variables = self._graph.get_all_nodes() \
             - set(treatment_name) \
-            - set(outcome_name) \
-            - set(self._graph.get_instruments(treatment_name, outcome_name))
+            - set(outcome_name)
         eligible_variables -= self._graph.get_descendants(treatment_name)
-
+        
         num_iterations = 0
         found_valid_adjustment_set = False
-        if self.method_name in CausalIdentifier.VALID_METHOD_NAMES:
-            for size_candidate_set in range(len(eligible_variables), 0, -1):
+        if method_name in CausalIdentifier.METHOD_NAMES:
+            # If `minimal-adjustment` method is specified, start the search from the set with minimum size. Otherwise, start from the largest.
+            set_sizes = range(1, len(eligible_variables) + 1, 1) if method_name == CausalIdentifier.BACKDOOR_MIN else range(len(eligible_variables), 0, -1)
+            for size_candidate_set in set_sizes:
                 for candidate_set in itertools.combinations(eligible_variables, size_candidate_set):
-                    check = self._graph.check_valid_backdoor_set(treatment_name,
-                            outcome_name, candidate_set, backdoor_paths=backdoor_paths)
-                    self.logger.debug("Candidate backdoor set: {0}, is_dseparated: {1}, No. of paths blocked by observed_nodes: {2}".format(candidate_set, check["is_dseparated"], check["num_paths_blocked_by_observed_nodes"]))
-                    if check["is_dseparated"]:
-                        backdoor_sets.append({
-                            'backdoor_set': candidate_set,
-                            'num_paths_blocked_by_observed_nodes': check["num_paths_blocked_by_observed_nodes"]})
-                        if self._graph.all_observed(candidate_set):
+                    # If include_unobserved is false, make sure the candidate set contains only observed variables.
+                    if include_unobserved or self._graph.all_observed(candidate_set):
```

Comment:
Can this done in a more efficient way by restricting the `eligible_variables` at the time of definition?
If `include_unobserved` is False, then we can check that eligible variables contains only observed variables and remove any unobserved variable. If `include_unobserved` is True, we do nothing.

This will avoid checking `_graph.all_observed` each time in the loop.

Reply:
Yeah, this sounds good. I remember thinking about it myself at one point, but for some reason I left it as is.

Reply:
This actually had a very elegant solution, a lot of the code was already there to be used :).